engines/sludge/savedata.cpp

```diff
@@ -0,0 +1,279 @@
+
+#define LOAD_ERROR "Can't load custom data...\n\n"
+
+namespace Sludge {
+
+unsigned short saveEncoding = false;
+char encode1 = 0;
+char encode2 = 0;
+
+extern char *gamePath;
+
+/*
+ void loadSaveDebug (char * com) {
+ FILE * ffpp = fopen ("debuggy.txt", "at");
+ fprintf (ffpp, "%s\n", com);
+ fclose (ffpp);
+ }
+
+ void loadSaveDebug (char com) {
+ FILE * ffpp = fopen ("debuggy.txt", "at");
+ fprintf (ffpp, "%c\n", com);
+ fclose (ffpp);
+ }
+
+ void loadSaveDebug (int com) {
+ FILE * ffpp = fopen ("debuggy.txt", "at");
+ fprintf (ffpp, "%d\n", com);
+ fclose (ffpp);
+ }
+ */
+
+void writeStringEncoded(const char *s, Common::WriteStream *stream) {
+	int a, len = strlen(s);
+
+	stream->writeUint16BE(len);
+	for (a = 0; a < len; a++) {
+		stream->writeByte(s[a] ^ encode1);
+		encode1 += encode2;
+	}
+}
+
+char *readStringEncoded(Common::File *fp) {
+	int a, len = fp->readUint16BE();
+	char *s = new char[len + 1];
+	if (!checkNew(s))
+		return NULL;
+	for (a = 0; a < len; a++) {
+		s[a] = (char)(fp->readByte() ^ encode1);
+		encode1 += encode2;
+	}
+	s[len] = 0;
+	return s;
+}
+
+char *readTextPlain(Common::File *fp) {
+	int32_t startPos;
+
+	uint32 stringSize = 0;
+	bool keepGoing = true;
+	char gotChar;
+	char *reply;
+
+	startPos = fp->pos();
+
+	while (keepGoing) {
+		gotChar = (char)fp->readByte();
+		if ((gotChar == '\n') || (fp->eos())) {
+			keepGoing = false;
+		} else {
+			stringSize++;
+		}
+	}
+
+	if ((stringSize == 0) && (fp->eos())) {
+		return NULL;
+	} else {
+		fp->seek(startPos, SEEK_SET);
+		reply = new char[stringSize + 1];
+		if (reply == NULL)
+			return NULL;
+		size_t bytes_read = fp->read(reply, stringSize);
+		if (bytes_read != stringSize && fp->err()) {
+			debugOut("Reading error in readTextPlain.\n");
+		}
+		fp->readByte();  // Skip the newline character
+		reply[stringSize] = 0;
+	}
+
+	return reply;
+}
+
+bool fileToStack(char *filename, stackHandler *sH) {
+
+	variable stringVar;
+	stringVar.varType = SVT_NULL;
+	const char *checker = saveEncoding ? "[Custom data (encoded)]\r\n" : "[Custom data (ASCII)]\n";
+
+	Common::File fd;
+
+	if (!fd.open(filename)) {
+#if 0
+		char currentDir[1000];
+		if (!getcwd(currentDir, 998)) {
+			debugOut("Can't get current directory.\n");
+		}
+
+		if (chdir(gamePath)) {
+			debugOut("Error: Failed changing to directory %s\n", gamePath);
+		}
+
+		if (chdir(currentDir)) {
+			debugOut("Error: Failed changing to directory %s\n", currentDir);
+		}
+
+		if (!fd.open(filename)) {
+			return fatal("No such file", filename);
+		}
+#endif
+		return fatal("No such file", filename); //TODO: false value
+	}
+
+	encode1 = (unsigned char)saveEncoding & 255;
+	encode2 = (unsigned char)(saveEncoding >> 8);
+
+	while (*checker) {
+		if (fd.readByte() != *checker) {
+			fd.close();
+			return fatal(LOAD_ERROR "This isn't a SLUDGE custom data file:", filename);
+		}
+		checker++;
+	}
+
+	if (saveEncoding) {
+		char *checker = readStringEncoded(&fd);
+		if (strcmp(checker, "UN�LO�CKED")) {
```

Comment:
here is UTF8, or something. If you have to use special characters, then replace them with their codes.